crates/accelerate/src/twirling.rs

```diff
@@ -0,0 +1,484 @@
+    ([ZGate, YGate, XGate, IGate], PI),
+    ([ZGate, IGate, IGate, ZGate], 0.),
+    ([ZGate, ZGate, ZGate, ZGate], 0.),
+    ([XGate, YGate, YGate, XGate], 0.),
+    ([XGate, IGate, ZGate, YGate], 0.),
+    ([XGate, ZGate, IGate, YGate], 0.),
+    ([XGate, XGate, XGate, XGate], 0.),
+    ([YGate, IGate, ZGate, XGate], PI),
+    ([YGate, ZGate, IGate, XGate], PI),
+    ([YGate, XGate, XGate, YGate], 0.),
+    ([YGate, YGate, YGate, YGate], 0.),
+];
+
+static TWIRLING_SETS: [&[([StandardGate; 4], f64); 16]; 4] = [
+    &CX_TWIRL_SET,
+    &CZ_TWIRL_SET,
+    &ECR_TWIRL_SET,
+    &ISWAP_TWIRL_SET,
+];
+
+const CX_MASK: u8 = 8;
+const CZ_MASK: u8 = 4;
+const ECR_MASK: u8 = 2;
+const ISWAP_MASK: u8 = 1;
+
+#[inline(always)]
+fn diff_frob_norm_sq(array: ArrayView2<Complex64>, gate_matrix: ArrayView2<Complex64>) -> f64 {
+    let mut res: f64 = 0.;
+    for i in 0..4 {
+        for j in 0..4 {
+            let gate = gate_matrix[[i, j]];
+            let twirled = array[[i, j]];
+            let diff = twirled - gate;
+            res += (diff.conj() * diff).re;
+        }
+    }
+    res
+}
+
+fn generate_twirling_set(gate_matrix: ArrayView2<Complex64>) -> Vec<([StandardGate; 4], f64)> {
+    let mut out_vec = Vec::with_capacity(16);
+    let i_matrix = aview2(&ONE_QUBIT_IDENTITY);
+    let x_matrix = aview2(&qiskit_circuit::gate_matrix::X_GATE);
+    let y_matrix = aview2(&qiskit_circuit::gate_matrix::Y_GATE);
+    let z_matrix = aview2(&qiskit_circuit::gate_matrix::Z_GATE);
+    let iter_set = [IGate, XGate, YGate, ZGate];
+    let kron_set: [Array2<Complex64>; 16] = [
+        kron(&i_matrix, &i_matrix),
+        kron(&x_matrix, &i_matrix),
+        kron(&y_matrix, &i_matrix),
+        kron(&z_matrix, &i_matrix),
+        kron(&i_matrix, &x_matrix),
+        kron(&x_matrix, &x_matrix),
+        kron(&y_matrix, &x_matrix),
+        kron(&z_matrix, &x_matrix),
+        kron(&i_matrix, &y_matrix),
+        kron(&x_matrix, &y_matrix),
+        kron(&y_matrix, &y_matrix),
+        kron(&z_matrix, &y_matrix),
+        kron(&i_matrix, &z_matrix),
+        kron(&x_matrix, &z_matrix),
+        kron(&y_matrix, &z_matrix),
+        kron(&z_matrix, &z_matrix),
+    ];
+    for (i_idx, i) in iter_set.iter().enumerate() {
+        for (j_idx, j) in iter_set.iter().enumerate() {
+            let before_matrix = kron_set[i_idx * 4 + j_idx].view();
+            let half_twirled_matrix = gate_matrix.dot(&before_matrix);
+            for (k_idx, k) in iter_set.iter().enumerate() {
+                for (l_idx, l) in iter_set.iter().enumerate() {
+                    let after_matrix = kron_set[k_idx * 4 + l_idx].view();
+                    let twirled_matrix = after_matrix.dot(&half_twirled_matrix);
+                    let norm: f64 = diff_frob_norm_sq(twirled_matrix.view(), gate_matrix);
+                    if norm.abs() < 1e-15 {
+                        out_vec.push(([*i, *j, *k, *l], 0.));
+                    } else if (norm - 16.).abs() < 1e-15 {
```

Comment:
This looks like something that could give a false positive, i.e. some custom defined gate where the difference is 16 yet they differ by a relative phase or something... or is that not possible?